# Numerical Instability that its source is in the SOLUTION method

### Question:
We will want to approximate the solution of this specific integral:

$$I=\int_0^1 x^7e^{x-1} \, dx $$

### Solution:
1. <b>Analytic calculation:</b>
    * can be computed directly by integration in parts integrations, BUT will need to repeat that 7 times...
    * One other simple way to calculate the integral above its via a defining recursic equation to calculate the integral:

$$
I=\int_0^1 x^ne^{x-1} \,dx,
n=0,1,2,\ldots
$$

By one <b>parts integration</b> will recieve:
$$
I=\int_0^1 x^ne^{x-1} \,dx = 
    \Bigg[ \begin{align*} 
        u=x^n \quad v'=e^{x-1}\\ 
        u'=nx^{n-1} \quad v=e^{x-1}
    \end{align*} 
\Bigg] =
\underbrace{x^n \cdot e^{x-1}\Bigr|_{0}^{1}}_\text{=1} -n \cdot
\underbrace{\int_0^1 x^ne^{x-1} \,dx}_\text{$I_{n-1}$} = 
1-n \cdot I_{n-1}
$$

So we have got the next recursic exuation:
$$
I_n = 1-n \cdot I_{n-1}, \quad n=0,1,2,\ldots
$$


Initialy condition $I_0$ can be calculated directly from: $I_0 = \int_0^1 e^{x-1}dx = e^{x-1}\Bigr|_{0}^{1} = 1-\frac{1}{e}$

The desired integral is infact $I_7$ in the defined recursive series hence by applying the recursive process 7 times will get:
$$
I_0 = 1 - \frac{1}{e}
$$


In [148]:
import numpy as np
numerical_result_ittr_0 = 1 - (1/np.exp(1))
mentisa_4_result_ittr_0 = round(numerical_result_ittr_0,4)

print ("I_0: {:33}".format(numerical_result_ittr_0))
print("I*_0_mentisa_4: {:10.4f}".format(mentisa_4_result_ittr_0))

I_0:                0.6321205588285577
I*_0_mentisa_4:     0.6321


$$
I_1 = 1- I_0 = \frac{1}{e}\\
$$

In [149]:
numerical_result_ittr_1 = 1 - numerical_result_ittr_0

mentisa_4_result_ittr_1 = 1 - mentisa_4_result_ittr_0
mentisa_4_result_ittr_1 = round(mentisa_4_result_ittr_1,4)

print ("I*_1: {:38}".format(numerical_result_ittr_1))
print("I_1_round_mentisa_4: {:10.4f}".format(mentisa_4_result_ittr_1))

I*_1:                    0.36787944117144233
I_1_round_mentisa_4:     0.3679


$$
I_2 = 1- 2 \cdot I_1 = 1 - \frac{2}{e}
$$

In [150]:
numerical_result_ittr_2 = 1 - 2*numerical_result_ittr_1

mentisa_4_result_ittr_2 = 1 - 2*mentisa_4_result_ittr_1
mentisa_4_result_ittr_2 = round(numerical_result_ittr_2,4)

print ("I_2: {:40}".format(numerical_result_ittr_2))
print("I*_2_round_mentisa_4: {:10.4f}".format(mentisa_4_result_ittr_2))

I_2:                      0.26424111765711533
I*_2_round_mentisa_4:     0.2642


$$
I_3 = 1- 3 \cdot I_2 = 1 - 3\cdot \Big( 1 - \frac{2}{e}\Big) = -2 + \frac{6}{e}
$$

In [151]:
numerical_result_ittr_3 = 1 - 3*numerical_result_ittr_2

mentisa_4_result_ittr_3 = 1 - 3*mentisa_4_result_ittr_2
mentisa_4_result_ittr_3 = round(mentisa_4_result_ittr_3,4)

print ("I_3: {:38}".format(numerical_result_ittr_3))
print("I*_3_round_mentisa_4: {:10.4f}".format(mentisa_4_result_ittr_3))

I_3:                      0.207276647028654
I*_3_round_mentisa_4:     0.2074


$$
I_4 = 1- 4 \cdot I_3 = 1 - 4\cdot \Big(  -2 + \frac{6}{e}\Big) = 9-\frac{24}{e}
$$

In [152]:
numerical_result_ittr_4 = 1 - 4*numerical_result_ittr_3

mentisa_4_result_ittr_4 = 1 - 4*mentisa_4_result_ittr_3
mentisa_4_result_ittr_4 = round(mentisa_4_result_ittr_4,4)

print ("I_4: {:40}".format(numerical_result_ittr_4))
print("I*_4_round_mentisa_4: {:10.4f}".format(mentisa_4_result_ittr_4))

I_4:                      0.17089341188538398
I*_4_round_mentisa_4:     0.1704


$$
I_5 = 1- 5 \cdot I_4 = 1 - 5\cdot \Big(  9-\frac{24}{e}\Big) = -44+\frac{120}{e}
$$

In [153]:
numerical_result_ittr_5 = 1 - 5*numerical_result_ittr_4

mentisa_4_result_ittr_5 = 1 - 5*mentisa_4_result_ittr_4
mentisa_4_result_ittr_5 = round(mentisa_4_result_ittr_5,4)

print ("I_5: {:41}".format(numerical_result_ittr_5))
print("I*_5*_round_mentisa_4: {:10.4f}".format(mentisa_4_result_ittr_5))

I_5:                       0.14553294057308008
I*_5*_round_mentisa_4:     0.1480


$$
I_6 = 1- 6 \cdot I_5 = 1 - 6\cdot \Big(-44+\frac{120}{e}\Big) = 255 + \frac{720}{e}
$$

In [154]:
numerical_result_ittr_6 = 1 - 6*numerical_result_ittr_5

mentisa_4_result_ittr_6 = 1 - 6*mentisa_4_result_ittr_5
mentisa_4_result_ittr_6 = round(mentisa_4_result_ittr_6,4)

print ("I_6: {:39}".format(numerical_result_ittr_6))
print("I*_6_round_mentisa_4: {:10.4f}".format(mentisa_4_result_ittr_6))

I_6:                      0.1268023565615195
I*_6_round_mentisa_4:     0.1120


$$
I_7 = 1- 7 \cdot I_6 = 1 - 7\cdot \Big(255 + \frac{720}{e}\Big) = -1854+\frac{5040}{e}
$$

In [155]:
numerical_result_ittr_7 = 1 - 7*numerical_result_ittr_6

mentisa_4_result_ittr_7 = 1 - 7*mentisa_4_result_ittr_6
mentisa_4_result_ittr_7 = round(mentisa_4_result_ittr_7,4)

print ("I_7: {:40}".format(numerical_result_ittr_7))
print("I*_7_round_mentisa_4: {:10.4f}".format(mentisa_4_result_ittr_7))

I_7:                      0.11238350406936348
I*_7_round_mentisa_4:     0.2160


Because of the limited digit number in the <b>MANTISE</b>, the recieved aproximate value is:
$$
0.1268023565615195 = I_7\cong I^*_7 = 0.2160 
$$

### Error calculations:
* <b>Absolute error</b> (approximated with 4 digits in the mantise):
    
    $$
    A.E = \bigr|I_7 - I^*_7\bigr| = \bigr|0.1124 - 0.2160\bigr| = 
    $$

In [156]:
a_e = round(abs(numerical_result_ittr_7 - mentisa_4_result_ittr_7),4)
print(a_e)

0.1036


* <b>Relative error in precent</b> (approximated with 4 digits in the mantise):    
$$
R.E = \frac{\bigr|I_7 - I^*_7\bigr|}{I_7}\times100\% = \frac{\bigr|0.1124 - 0.2160\bigr|}{0.1124}\times100\%=
$$

In [157]:
r_e = abs((round(numerical_result_ittr_7,4) - mentisa_4_result_ittr_7)/round(numerical_result_ittr_7,4))*100
r_e = round(r_e,2)
print("{}%".format(r_e))

92.17%


This big error inticate for numeric instability.

### Source / Magnitude of the error
will try to understand the source of the error, and evaluate in <b>advance</b> what wil be the error magnitude.

Already in the pre-iterative stage, we have calculated an approximation of $I_0$ rounded with 4 digits in the mantice. Hence errors magnitude in this stage (with 4 digits in mantice) wil be:
$$
err_0 = \bigr|I_0-I^*_0\bigr| = \big(1-\frac{1}{e}\big) - 0.6321\bigr| = 
$$

In [158]:

err_0 = numerical_result_ittr_0 - mentisa_4_result_ittr_0
print("{:0.10f}".format(err_0))

0.0000205588


# Results
Show graphs and stats here

# Conclusions and Next Steps
Summarize findings here